CMakeLists.txt

```diff
@@ -0,0 +1,273 @@
+endif ()
+
+# other packages
+find_package (Doxygen)
+find_package (BMV2)
+
+# enable CTest
+enable_testing ()
+
+
+# if we want to manage versions in CMake ...
+# include (cmake/P4CVersion.cmake)
+# set (CPACK_PACKAGE_VERSION_MAJOR ${__P4C_VERSION_MAJOR})
+# set (CPACK_PACKAGE_VERSION_MINOR ${__P4C_VERSION_MINOR})
+# set (CPACK_PACKAGE_VERSION_PATCH ${__P4C_VERSION_PATCH})
+# if (__P4C_VERSION_RC)
+#   set (CPACK_PACKAGE_VERSION_PATCH ${CPACK_PACKAGE_VERSION_PATCH}-${__P4C_VERSION_RC})
+# endif ()
+
+# set (CMAKE_CXX_EXTENSIONS OFF) # prefer using -std=c++11 rather than -std=gnu++11
+set (CMAKE_CXX_STANDARD 11)
+set (CMAKE_CXX_STANDARD_REQUIRED ON)
+
+include(P4CUtils)
+
+add_cxx_compiler_option ("-Wall")
+add_cxx_compiler_option ("-Wextra")
+add_cxx_compiler_option ("-Wno-overloaded-virtual")
+add_cxx_compiler_option ("-Wno-deprecated")
+
+include_directories (
+  ${P4C_SOURCE_DIR}/frontends
+  ${P4C_SOURCE_DIR}/backends
+  ${P4C_SOURCE_DIR}/extensions
+  ${P4C_SOURCE_DIR}
+  ${P4C_SOURCE_DIR}/test/frameworks/gtest/googletest/include
+  ${P4C_BINARY_DIR}
+  )
+add_definitions (-DCONFIG_PREFIX="${CMAKE_INSTALL_PREFIX}")
+add_definitions (-DCONFIG_PKGDATADIR="${P4C_ARTIFACTS_OUTPUT_DIRECTORY}")
+
+set (CMAKE_CXX_FLAGS         "${CMAKE_CXX_FLAGS} ${P4C_CXX_FLAGS}")
+set (CMAKE_CXX_FLAGS_DEBUG   "${CMAKE_CXX_FLAGS_DEBUG} ${P4C_CXX_FLAGS}")
+set (CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} ${P4C_CXX_FLAGS}")
+set (CPPLINT_FILES)          # list to collect all files that need lint
+set (TEST_TAGS "p4" CACHE INTERNAL "test tags") # list to collect all test tags
+set (IR_DEF_FILES)           # list to collect all .def files
+
+# Other configuration files that need to be generated
+configure_file ("${P4C_SOURCE_DIR}/cmake/config.h.cmake" "${P4C_BINARY_DIR}/config.h")
+
+set (IR_GENERATED_SRCS
+  ${P4C_BINARY_DIR}/ir/ir-generated.h
+  ${P4C_BINARY_DIR}/ir/ir-generated.cpp
+  ${P4C_BINARY_DIR}/ir/gen-tree-macro.h)
+set (IR_GENERATOR ${P4C_BINARY_DIR}/tools/ir-generator/irgenerator)
+
+# the order of adding subdirectories matters because of target dependencies
+add_subdirectory (tools/driver)
+add_subdirectory (lib)
+add_cpplint_files(lib "${LIB_CPPLINT_FILES}")
+add_subdirectory (tools/ir-generator)
+add_cpplint_files(tools/ir-generator "${IRGEN_CPPLINT_FILES}")
+add_subdirectory (ir)
+add_cpplint_files(ir "${IR_CPPLINT_FILES}")
+
+# component libraries: must be defined before being used in the
+# backend executables
+set (P4C_LIBRARIES controlplane midend frontend ir p4ctoolkit)
+
+# add extensions - before the frontends as they produce IR and extra frontend
+set(EXTENSION_FRONTEND_SOURCES)
+file (GLOB p4c_extensions RELATIVE ${P4C_SOURCE_DIR}/extensions ${P4C_SOURCE_DIR}/extensions/*)
+MESSAGE ("-- Available extensions ${p4c_extensions}")
+foreach (ext ${p4c_extensions})
+  if (EXISTS ${P4C_SOURCE_DIR}/extensions/${ext}/CMakeLists.txt)
+    add_subdirectory (extensions/${ext})
+    set (__cpplint_list "${${ext}_CPPLINT_FILES}")
```

Comment:
There's a lot of gross machinery around cpplint in here, and unfortunately we will need similar stuff for some other cases as well. I wonder if we shouldn't just use the idiom we're using for `IR_DEF_FILES` for cpplint too. Something even cleaner might be to use source file properties, but I'm not sure if there are scoping issues there that would prevent us from using them.